source/Font.cpp

```diff
@@ -12,91 +12,169 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
-		"void main() {\n"
-		"  texCoord = vec2((glyph + corner.x) / 98.f, corner.y);\n"
-		"  gl_Position = vec4((aspect * vert.x + position.x) * scale.x, (vert.y + position.y) * scale.y, 0, 1);\n"
-		"}\n";
-	
-	const char *fragmentCode =
-		// The user must supply a texture and a color (white by default).
-		"uniform sampler2D tex;\n"
-		"uniform vec4 color = vec4(1, 1, 1, 1);\n"
+		// 2 bytes - 110xxxxx 10xxxxxx
+		if((*str & 0x20) == 0)
+			return 2;
 		
-		// This comes from the vertex shader.
-		"in vec2 texCoord;\n"
+		// invalid - 111????? 10?????? invalid
+		if((*(str + 2) & 0xc0) != 0x80)
+			return -1;
 		
-		// Output color.
-		"out vec4 finalColor;\n"
+		// 3 bytes - 1110xxxx 10xxxxxx 10xxxxxx
+		if((*str & 0x10) == 0)
+			return 3;
 		
-		// Multiply the texture by the user-specified color (including alpha).
-		"void main() {\n"
-		"  finalColor = texture(tex, texCoord).a * color;\n"
-		"}\n";
-	
-	const int KERN = 2;
+		// invalid - 1111???? 10?????? 10?????? invalid
+		if((*(str + 3) & 0xc0) != 0x80)
+			return -1;
+		
+		// 4 bytes - 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+		if((*str & 0x8) == 0)
+			return 4;
+		
+		// not unicode - 11111??? 10?????? 10?????? 10??????
+		return -1;
+	}
 }
 
 
 
 Font::Font()
-	: texture(0), vao(0), vbo(0), colorI(0), scaleI(0), glyphI(0), aspectI(0),
-	  positionI(0), height(0), space(0), screenWidth(0), screenHeight(0)
+	: size(0)
 {
 }
 
 
 
-Font::Font(const string &imagePath)
-	: Font()
+bool Font::Load(const DataNode &node)
 {
-	Load(imagePath);
+	// Ignore if already loaded.
+	if(!sources.empty())
+		return false;
+	
+	if(node.Size() < 1 || node.Token(0) != "font")
+	{
+		node.PrintTrace("was expecting font");
+		return false;
+	}
+	
+	// Get size.
+	size = 0;
+	for(const DataNode &child : node)
+	{
+		string key = (child.Size() >= 0 ? child.Token(0) : "");
+		if(key != "size")
+			continue;
+		
+		if(size > 0)
+		{
+			node.PrintTrace("too many sizes");
+			return false;
+		}
+		size = round(child.Value(1));
+		if(size <= 0)
+		{
+			child.PrintTrace("invalid font.size");
+			return false;
+		}
+	}
+	if(size <= 0)
+	{
+		node.PrintTrace("missing size");
+		return false;
+	}
+	
+	// Get glyph source.
+	sources.clear();
+	for(const DataNode &child : node)
+	{
+		if(child.Size() <= 0 || (child.Token(0) != "atlas" && child.Token(0) != "freetype"))
+			continue;
+		string key = child.Token(0);
+		
+		if(child.Size() <= 1 || child.Token(1).empty())
+		{
+			child.PrintTrace("missing path");
+			return false;
+		}
+		
+		vector<string> paths;
+		paths.emplace_back(child.Token(1));
+		paths.emplace_back(Files::Resources() + child.Token(1));
+		for(const string &source : GameData::Sources())
+			paths.emplace_back(source + child.Token(1));
+		while(!paths.empty() && !Files::Exists(paths.back()))
+			paths.pop_back();
+		if(paths.empty())
+		{
+			child.PrintTrace("path not found");
+			return false;
+		}
+		
+		size_t count = sources.size();
+		if(key == "atlas")
```

Comment:
I wonder if you can use the implicit constructors to simplify this even further, i.e. do away with checking count & sources.size()
```
sources.emplace_back(paths.back())
if(!sources.back().IsLoaded())
    child.PrintTrace("Load failed for valid path '" + paths.back() + "':");
```
It probably can't simplify to that extent, but surely a check to sources.back().IsLoaded() could be implemented.

Reply:
`sources` should only contain valid interfaces.

Concrete classes can have different load mechanics, like AtlasGlyphs not needing the size.
When new features are introduced to a type of glyph source, it will probably receive the child DataNode so it can read the grandchildren.

Reply:
Yeah, i left off the `.pop_back()` and other error handling cleanup from the example, since the example wasn't likely to be implementable, but served well to convey the intent.